Clear credentials from clipboard after 60 seconds

FabianHenneke opened the pull request:
My implementation for #140. This works well on Chrome, but Firefox somehow hangs on the following line indefinitely:
```js
const alarms = await chrome.alarms.getAll();
```
I don't have any experience with Firefox extensions, do you know why this may happen?

----

Fixes #140

maximbaz commented on src/background.js:
```diff
@@ -144,6 +173,36 @@ function copyToClipboard(text) {
     );
     document.execCommand("copy");
+
+    if (!doNotClear) {
```
there's too much negation for my taste, let's rename the parameter to `clear = true`, and this becomes `if (clear)`

maximbaz commented on src/manifest-firefox.json:
```diff
@@ -26,7 +26,9 @@
         "activeTab",
+        "alarms",
         "tabs",
+        "clipboardRead",
```
Please add these to README in the section describing permissions

erayd replied:
@maximbaz We should maybe do this in 3.2, because adding this permission will disable it for everybody when the extension auto-updates.

maximbaz replied:
Good point - let's target this for 3.2

@FabianHenneke please update `@since 3.1.2` to `@since 3.2.0` :)

maximbaz commented on src/manifest-firefox.json:
```diff
@@ -26,7 +26,9 @@
+        "alarms",
         "tabs",
+        "clipboardRead",
         "clipboardWrite",
```
Let's also add to README a note saying that copied credentials will be cleared in 60 seconds.

Nothing fancy, perhaps just update "Available keyboard shortcuts" section:

> <kbd>Ctrl+C</kbd> Copy password to clipboard (will clear in 60 seconds)
> <kbd>Ctrl+Shift+C</kbd> Copy username to clipboard (will clear in 60 seconds)

maximbaz commented on src/background.js:
```diff
@@ -80,6 +80,18 @@ chrome.commands.onCommand.addListener(async command => {
+// handle fired alarms
+chrome.alarms.onAlarm.addListener(async alarm => {
+    if (alarm.name.startsWith("clearClipboard-")) {
+        const fingerprint = await calculateFingerprint(readFromClipboard());
```
Because it seems we are guaranteed to always have either 0 or 1 active alarm to clear clipboard, how about we hardcode the alarm name as "clear-clipboard", and then in some global variable keep either `null` or last copied data? This way we don't really need to hash anything as the data stays local to background script, and here you would have something like 

```
if (alarm.name === "clear-clipboard" && readFromClipboard() === lastCopiedData) {
    await copyToClipboard(" ", true);
}
lastCopiedData = null
```

I hope alarms are scoped to our extension, i.e. no other extension is able to read or trigger our alarms, do you know?

FabianHenneke replied:
My idea here was to minimize our dependency on the background page being persistent, which will likely go away in manifest version 3. If you think that it's fine to add features using persistence in the meantime, than I would also gladly go for the easier solution.

maximbaz replied:
Let's go for easier solution, given that we already have persistence for modal HTTP auth and I'm gonna add one more for badge counters to properly solve #135, it's not going away any time soon 🙂

maximbaz commented on src/background.js:
```diff
@@ -144,6 +173,36 @@ function copyToClipboard(text) {
+        }
+        // clear clipboard after 1 minute if its content still has the same fingerprint as text
+        const fingerprint = await calculateFingerprint(text);
+        await chrome.alarms.create(`clearClipboard-${fingerprint}`, { delayInMinutes: 1 });
```
I'm not sure this function returns a Promise, and therefore you probably don't need to await it, and this entire function doesn't need to be `async`. Could you confirm?

By the way, you don't need to clear the alarm beforehand:

> Alarm names are unique within the scope of a single extension. If an alarm with an identical name exists, the existing alarm will be cleared and the alarm being created will replace it.

So all you need here is:

```
if (clear) {
    chrome.alarms.create("clear-clipboard", { delayInMinutes: 1 });
    lastCopiedData = text;
}
```

maximbaz commented on src/background.js:
```diff
@@ -80,6 +80,18 @@ chrome.commands.onCommand.addListener(async command => {
+        // clear the clipboard if the fingerprint of the content hasn't changed
+        if (alarm.name === `clearClipboard-${fingerprint}`) {
+            // setting clipboard to "" is not fully reliable, hence set it to " "
+            await copyToClipboard(" ", true);
```
Why do you think it's not reliable? Works well for me in Chromium and Firefox, plus it doesn't pollute clipboard with a space character, but simply clears it.

FabianHenneke replied:
It was not reliable when I first implemented it for smart-pass, but that is about two years ago. If it works, it's clearly preferable to copying a space character.

FabianHenneke commented:
Thanks for the detailed review! I have now addressed all the comments.
It took me a while to figure out the textarea incantations that please both Firefox and Chrome when it comes to pasting clipboard contents, but it seems to work reliably in both browsers now. Could you also verify this?

FabianHenneke commented:
For the record: https://bugzilla.mozilla.org/show_bug.cgi?id=1364708#c2 is the only code snippet that I found that works.

maximbaz commented on README.md:
```diff
@@ -272,6 +272,7 @@ Browserpass extension requests the following permissions:
 | `debugger`           | Only used for "auto-submit" functionality: if all attepts to locate a "Submit" button failed, Browserpass will put focus inside the login form and issue an <kbd>Enter</kbd> keypress. This is only available in Chromium-based browsers, and sadly this permission [cannot be specified as optional](https://developer.chrome.com/apps/permissions) |
 | `activeTab`          | To get URL of the current tab, used for example to determine which passwords to show you by default in the popup                                                                                                                                                                                                                                     |
 | `tabs`               | To get URL of a given tab, used for example to set count of the matching passwords for a given tab                                                                                                                                                                                                                                                   |
+| `clipboardRead`      | To ensure only copied credentials and not other content is cleared from the clipboard after 60 seconds                                                                                                                                                                                                                                         |
```
Please also add a note about the `alarms` permission

maximbaz commented on src/background.js:
```diff
@@ -144,6 +158,31 @@ function copyToClipboard(text) {
+
+    if (clear) {
+        lastCopiedText = text;
+        chrome.alarms.create(`clearClipboard`, { delayInMinutes: 1 });
```
you can change back to regular `"` string here

erayd commented:
Related to the permissions discussion above - can the `clipboardRead` permission be requested on demand the first time somebody tries to use this feature? That would help avoid a mass-disabling of the extension for users, and would help to clarify its purpose for those who don't read the documentation (which let's face it, will be a large fraction of the user base).

maximbaz commented:
I'm not sure that's a good idea to add optional permissions at this point, it adds a bit of complexity, potential browser issues, and inconsistency with other permissions that are not optional...

How about we keep them required and instead add an upgrade notification "New permissions added to clear copied credentials after 60 seconds"? We already have a framework for that in `background.js#onExtensionInstalled()`.

FabianHenneke commented:
Would this work though? I think that the extension would be disabled before the notification can be shown. But of course this would at least explain why after the fact better than the permission dialog by itself.

maximbaz commented:
It will be disabled for those who install extension via webstore, but as soon as people re-enable it they will instantly understand why we suddenly requested more permissions — sounds acceptable to me, a good compromise between simplicity and usability, let's ask what @erayd thinks 🙂

FabianHenneke commented:
I implemented @maximbaz's notification proposal in the newest changeset. @erayd, if you would prefer the permission to be optional and know how to make this work in Firefox, I could always revert to the previous version.

erayd commented:
@FabianHenneke If you aren't able to get optional permissions to behave in Firefox, then let's leave things as-is, with the permission mandatory. Firefox is annoying enough already with its sporadic poor / incomplete API implementations, and this is probably just another such area.

FabianHenneke commented:
I haven't been able to get optional permissions to work in Firefox, so I would prefer to go with the notification approach.
